receiver/snmpreceiver/config_test.go

```diff
@@ -0,0 +1,1636 @@
+				assert.ErrorContains(t, err, test.expectedErr)
+			} else {
+				assert.NoError(t, err)
+			}
+			compareCfgs(t, test.expectedCfg, snmpConfig)
+		})
+	}
+}
+
+// Testing Validate directly to test that missing data errors when no defaults are provided
+func TestValidate(t *testing.T) {
+	type testCase struct {
+		name        string
+		cfg         *Config
+		expectedErr string
+	}
+
+	testCases := []testCase{
+		{
+			name: "NoEndpointErrors",
+			cfg: &Config{
+				Version:   "v2c",
+				Community: "public",
+				Metrics: map[string]MetricConfig{
+					"m3": {
+						Unit: "By",
+						Gauge: &GaugeMetric{
+							ValueType: "float",
+						},
+						ScalarOIDs: []ScalarOID{
+							{
+								OID: "1",
+							},
+						},
+					},
+				},
+			},
+			expectedErr: errEmptyEndpoint.Error(),
+		},
+		{
+			name: "NoVersionErrors",
+			cfg: &Config{
+				Endpoint:  "udp://localhost:161",
+				Community: "public",
+				Metrics: map[string]MetricConfig{
+					"m3": {
+						Unit: "By",
+						Gauge: &GaugeMetric{
+							ValueType: "float",
+						},
+						ScalarOIDs: []ScalarOID{
+							{
+								OID: "1",
+							},
+						},
+					},
+				},
+			},
+			expectedErr: errEmptyVersion.Error(),
+		},
+		{
+			name: "V3NoSecurityLevelErrors",
+			cfg: &Config{
+				Endpoint: "udp://localhost:161",
+				Version:  "v3",
+				User:     "u",
+				Metrics: map[string]MetricConfig{
+					"m3": {
+						Unit: "By",
+						Gauge: &GaugeMetric{
+							ValueType: "float",
+						},
+						ScalarOIDs: []ScalarOID{
+							{
+								OID: "1",
+							},
+						},
+					},
+				},
+			},
+			expectedErr: errEmptySecurityLevel.Error(),
+		},
+		{
+			name: "V3NoAuthTypeErrors",
+			cfg: &Config{
+				Endpoint:      "udp://localhost:161",
+				Version:       "v3",
+				SecurityLevel: "auth_no_priv",
+				User:          "u",
+				AuthPassword:  "p",
+				Metrics: map[string]MetricConfig{
+					"m3": {
+						Unit: "By",
+						Gauge: &GaugeMetric{
+							ValueType: "float",
+						},
+						ScalarOIDs: []ScalarOID{
+							{
+								OID: "1",
+							},
+						},
+					},
+				},
+			},
+			expectedErr: errEmptyAuthType.Error(),
+		},
+		{
+			name: "V3NoPrivacyTypeErrors",
+			cfg: &Config{
+				Endpoint:        "udp://localhost:161",
+				Version:         "v3",
+				SecurityLevel:   "auth_priv",
+				User:            "u",
+				AuthType:        "md5",
+				AuthPassword:    "p",
+				PrivacyPassword: "pp",
+				Metrics: map[string]MetricConfig{
+					"m3": {
+						Unit: "By",
+						Gauge: &GaugeMetric{
+							ValueType: "float",
+						},
+						ScalarOIDs: []ScalarOID{
+							{
+								OID: "1",
+							},
+						},
+					},
+				},
+			},
+			expectedErr: errEmptyPrivacyType.Error(),
+		},
+	}
+
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.cfg.Validate()
+			assert.ErrorContains(t, err, test.expectedErr)
+		})
+	}
+}
+
+func compareCfgs(t assert.TestingT, expected *Config, actual *Config) {
```

Comment:
I'm not sure I've seen any other configs that required custom comparison functions. Is this necessary?

Reply:
@djaglowski i could just manually do the code inside of here in a few different spots, but i specifically refactored it to be in this method so it's in one spot and not repeated. this was purely to keep things DRY

Reply:
I think you misunderstood my question. What I am trying to figure out, is why it is necessary in the first place to make comparisons of each field, as opposed to what it done elsewhere in the collector where we just have `require.Equal(t, expected, cfg)`. ([example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7429ab6db5db8a45926f652c22b0d32e4b2dc278/receiver/couchdbreceiver/config_test.go#L123))

Reply:
@djaglowski ah sorry. after redoing this with the `UnmarshalReceiver` like you suggested, I believe this will go away.